applications/convection_diffusion_application/python_scripts/adjoint_diffusion_solver.py

```diff
@@ -0,0 +1,259 @@
+        if self.solver_imports_model_part:
+            self._ImportModelPart(self.model_part,self.settings["model_import_settings"])
+
+    def PrepareModelPart(self):
+        if self.solver_imports_model_part:
+            # ensure that the element type is the correct one
+            self._set_elements_and_conditions()
+
+            # check mesh orientation (tetrahedral mesh orientation check)
+            throw_errors = False
+            kratos.TetrahedralMeshOrientationCheck(self.model_part, throw_errors).Execute()
+
+            # set the buffer size
+            if self.model_part.GetBufferSize() < self.min_buffer_size:
+                self.model_part.SetBufferSize(self.min_buffer_size)
+
+            # initialize the adjoint model part using primal results
+            primal_model_part = self.model.GetModelPart(self.primal_model_part_name)
+            variable_utils = kratos.VariableUtils()
+            variable_utils.CopyModelPartNodalVar(kratos.TEMPERATURE, primal_model_part, self.model_part, 0)
+            variable_utils.CopyModelPartNodalVar(kratos.FACE_HEAT_FLUX, primal_model_part, self.model_part, 0)
+
+            self.ImportMaterials()
+
+    def ImportMaterials(self):
+        materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
+        if (materials_filename != ""):
+            with open(materials_filename, 'r') as parameter_file:
+                materials = kratos.Parameters(parameter_file.read())
+
+            for i in range(materials["properties"].size()):
+                model_part = self.model.GetModelPart(materials["properties"][i]["model_part_name"].GetString())
+                mat = materials["properties"][i]["Material"]
+                var_utils = kratos.VariableUtils()
+                for key, value in mat["Variables"].items():
+                    var = kratos.KratosGlobals.GetVariable(key)
+                    #if not model_part.HasNodalSolutionStepVariable(var):
+                    #     raise Exception("Trying to set variable {0} on nodes, but the variable is not in nodal data.".format(var.Name()))
+                    if model_part.HasNodalSolutionStepVariable(var):
+                        if value.IsDouble():
+                            var_utils.SetScalarVar(var, value.GetDouble(), model_part.Nodes)
+                        elif value.IsVector():
+                            var_utils.SetVectorVar(var, value.GetVector(), model_part.Nodes)
+                        else:
+                            raise ValueError("Type of value is not available")
+                    else:
+                        kratos.Logger.PrintWarning("Ignoring variable {0} given by the materials file, since it is not a nodal variable used by this solver.".format(var.Name()))
+
+    def DefineConvectionDiffusionSettings(self,settings):
+        convection_diffusion_settings = kratos.ConvectionDiffusionSettings()
```

Comment:
Why is convection_diffusion_settings.h in kratos/includes?

Reply:
In spite of name, it is used by multiple applications besides the convection_diffusion_application (ThermoMechanical, FluidTransport and a couple more).